frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-builder/switch-to-form-validation-utils.ts

```diff
@@ -0,0 +1,110 @@
+import { PipelineTask } from '../../../types';
+import { PipelineBuilderFormValues } from './types';
+import { runAfterMatches } from './validation-utils';
+
+const resourceDefinitionYAML = () => {
+  return yup.array().of(
+    yup.object({
+      name: yup.string().required(),
+      resource: yup.string(),
+    }),
+  );
+};
+
+export const validRunAfter = (formData: PipelineBuilderFormValues, thisTask: PipelineTask) => {
+  return yup.array().of(
+    yup
+      .string()
+      .test('tasks-matches-runAfters', i18n.t('pipelines-plugin~Invalid runAfter'), function(
+        runAfter: string,
+      ) {
+        return runAfterMatches(formData, [runAfter], thisTask.name);
+      }),
+  );
+};
+
+const taskValidationYAMLSchema = (formData: PipelineBuilderFormValues) => {
+  return yup.array().of(
+    yup.lazy((taskObject) =>
+      yup
+        .object({
+          name: nameValidationSchema((tKey) => i18n.t(tKey)),
+          taskRef: yup
+            .object({
+              name: yup.string(),
+              kind: yup.string(),
+            })
+            .default(undefined),
+          taskSpec: yup.object(),
+          runAfter: validRunAfter(formData, taskObject),
+          params: yup.array().of(
+            yup.object({
+              name: yup.string().required(),
+              value: yup.lazy((value) => {
+                if (Array.isArray(value)) {
+                  return yup.array().of(yup.string());
+                }
+                return yup.string();
+              }),
+            }),
+          ),
+          resources: yup.object({
+            inputs: resourceDefinitionYAML(),
+            outputs: resourceDefinitionYAML(),
+          }),
+          when: yup.array().of(
+            yup.object({
+              input: yup.string(),
+              operator: yup.string(),
+              values: yup.array().of(yup.string()),
+            }),
+          ),
+
+          workspaces: yup.array().of(
+            yup.object({
+              name: yup.string().required(),
+              workspace: yup.string(),
+            }),
+          ),
+        })
+        .test(
+          'taskRef-or-taskSpec',
+          i18n.t('pipelines-plugin~TaskSpec or TaskRef must be provided.'),
+          function(task: PipelineTask) {
+            return !!task.taskRef || !!task.taskSpec;
+          },
+        ),
+    ),
+  );
+};
+
+export const pipelineBuilderYAMLSchema = (formData: PipelineBuilderFormValues) => {
+  return yup.object({
+    metadata: yup.object({ name: yup.string() }),
+    spec: yup.object({
+      params: yup.array().of(
+        yup.object({
+          name: yup.string(),
+          description: yup.string(),
+          default: yup.string(),
+        }),
+      ),
+      resources: yup.array().of(
+        yup.object({
+          name: yup.string(),
+          type: yup.string(),
+        }),
+      ),
+      workspaces: yup.array().of(
+        yup.object({
+          name: yup.string(),
+        }),
```

Comment:
I still see errors for the top level keys in the pipeline.spec, the required fields are not added in here.

Reply:
I can't seem to reproduce a similar scenario for these fields since they can be modified using the form.
could you please help me with the steps if I missed something here.

Reply:
Workspace, resources, params on the pipeline.spec has `name` field, if we make them required(), we could auto trim the invalid entries like below, do we want to do this?
```
spec:
  workspace:
    - test: workspace
```

Reply:
Ah I figured :facepalm: , adding .required() in the name will not allow this yaml to switch to form view. I was wrong

```
  workspaces:
    - name: ''
      optional: false
 ```